BitwardenShared/Core/Platform/Services/AuthenticatorSyncService.swift

```diff
@@ -0,0 +1,237 @@
+                id: cipher.id ?? UUID().uuidString,
+                name: cipher.name,
+                totpKey: cipher.login?.totp,
+                username: username
+            )
+        }
+    }
+
+    /// This function handles the initial syncing with the Authenticator app as well as listening for updates
+    /// when the user adds new items. This is called when the sync is turned on.
+    ///
+    /// - Parameter userId: The userId of the user who has turned on sync.
+    ///
+    private func handleSyncOnForUserId(_ userId: String) async {
+        guard !vaultTimeoutService.isLocked(userId: userId) else {
+            return
+        }
+
+        do {
+            try await createAuthenticatorKeyIfNeeded()
+        } catch {
+            errorReporter.log(error: error)
+        }
+        subscribeToCipherUpdates(userId: userId)
+    }
+
+    /// This function handles stopping sync and cleaning up all sync-related items when a user has turned sync Off.
+    ///
+    /// - Parameter userId: The userId of the user who has turned off sync.
+    ///
+    private func handleSyncOffForUserId(_ userId: String) {
+        cipherPublisherTasks[userId]??.cancel()
+        cipherPublisherTasks[userId] = nil
+    }
+
+    /// Subscribe to NotificationCenter updates about if the app is in the foreground vs. background.
+    ///
+    private func subscribeToAppState() {
+        Task {
+            for await _ in notificationCenterService.willEnterForegroundPublisher() {
+                subscribeToSyncToAuthenticatorSetting()
+            }
+        }
+    }
+
+    /// Create a task for the given userId to listen for Cipher updates and sync to the Authenticator store.
+    ///
+    /// - Parameter userId: The userId of the account to listen for.
+    ///
+    private func subscribeToCipherUpdates(userId: String) {
+        guard cipherPublisherTasks[userId] == nil else { return }
+
+        cipherPublisherTasks[userId] = Task {
+            do {
+                for try await ciphers in try await self.cipherService.ciphersPublisher().values {
+                    try await writeCiphers(ciphers: ciphers, userId: userId)
+                }
+            } catch {
+                errorReporter.log(error: error)
+            }
+        }
+    }
+
+    /// Subscribe to the Sync to Authenticator setting to handle when the user grants (or revokes)
+    /// permission to sync items to the Authenticator app.
+    ///
+    private func subscribeToSyncToAuthenticatorSetting() {
+        syncSettingSubscriberTask?.cancel()
+        syncSettingSubscriberTask = Task {
+            for await (userId, shouldSync) in await self.stateService.syncToAuthenticatorPublisher().values {
+                guard let userId else { continue }
+
+                if shouldSync {
+                    await handleSyncOnForUserId(userId)
+                } else {
+                    handleSyncOffForUserId(userId)
+                }
+            }
```

Comment:
🤔 I think we should take into consideration whether the vault is `locked` / `unlocked` to call the handle methods. I see this only happening inside the `handleSyncOn...` func but what would that work correctly in the next scenario?
1. The active user id is unlocked and should sync so `handleSyncOnForUserId` is called and it subscribes to cipher updates.
2. Then the active user id gets locked but as far as I see the cipher updates are not unsubscribed possibly introducing race conditions errors depending if the lock happens while decrypting totps or replacing the items.

IMO if the vault locks/logsout then it should cancel any running operations to avoid any unwanted behaviors.
What do you think?

Reply:
My intention with this PR was just to do the first minimalist pass, but Vault unlocking and background stuff is definitely on my mind. 👍  I'm working on that in a separate branch, I just wanted to check in with this PR first on the overall architecture of adding this service, proving out the basic sync, etc. As such, I was just bombing out if the Vault was unlocked. For this PR, you had to have an unlocked vault when you turn sync on, or background/foreground the app with sync on and vault unlocked for anything to work.

My current thinking is the best way is to listen to two publishers: the sync setting AND the vault lock/unlock. If we get a vault lock/unlock or if the user turns on/off sync, we need to re-evaluate the sync in light of both and clean up accordingly (e.g. the point you mentioned about the vault locking _while_ we were decrypting ciphers is an excellent one). What do you think about that approach?

Reply:
Oh excellent then and I agree that's a nice approach to take 😄 Thanks for clarifying.